runtime/callback.c

```diff
@@ -97,33 +137,79 @@ CAMLexport value caml_callback3_exn(value closure,
   return caml_callbackN_exn(closure, 3, arg);
 }
 
-#else
+#else /* Nativecode callbacks */
 
-/* Native-code callbacks. */
+static void init_callback_code(void)
+{
+}
 
-typedef value (callback_stub)(caml_domain_state* state, value closure,
+typedef value (callback_stub)(caml_domain_state* state,
+                              value closure,
                               value* args);
 
 callback_stub caml_callback_asm, caml_callback2_asm, caml_callback3_asm;
 
 CAMLexport value caml_callback_exn(value closure, value arg)
 {
-  return caml_callback_asm(Caml_state, closure, &arg);
+  caml_domain_state* domain_state = Caml_state;
+  caml_maybe_expand_stack();
+
+  if (Stack_parent(domain_state->current_stack)) {
+    CAMLparam2 (closure, arg);
+    CAMLlocal1 (cont);
+    value res;
+
+    cont = save_and_clear_stack_parent(domain_state);
+    res = caml_callback_asm(domain_state, closure, &arg);
+    restore_stack_parent(domain_state, cont);
+
+    CAMLreturn (res);
+  }
+  return caml_callback_asm(domain_state, closure, &arg);
```

Comment:
I think the control flow would be clearer like this:
```suggestion
  }else{
    return caml_callback_asm(domain_state, closure, &arg);
  }
```

Reply:
Fixed in 54365f2